File: solutions/leetcode/maximum-subarray.cpp
/* Problem maximum-subarray
 * https://www.leetcode.com/problems/maximum-subarray
 */

#include <vector>

using namespace std;

class Solution {
 public:
  int maxSubArray(vector<int>& nums) {
    int sol = nums[0], par = nums[0];
    for (int i = 1; i < nums.size(); ++i) {
      par = max(par, 0) + nums[i];
      sol = max(sol, par);
    }
    return sol;
  }
};

File: solutions/leetcode/maximum-subarray.rs
/* Problem maximum-subarray
 * https://www.leetcode.com/problems/maximum-subarray
 */

struct Solution;

impl Solution {
	pub fn max_sub_array(mut nums: Vec<i32>) -> i32 {
		let mut max = i32::MIN;
		let mut max_cur = 0;
		for num in nums {
			max_cur = i32::wrapping_add(i32::max(max_cur, 0), num);
			max = i32::max(max, max_cur);
		}
		max
	}
}

fn main() {}
